File: Go/5 kyu/Simple Fun #178 Faulty Odometer/kata.go
package kata

import (
	"fmt"
	"math"
	"strings"
)

func Reverse(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}

func PowInt(x, y int) int {
	return int(math.Pow(float64(x), float64(y)))
}

func FaultyOdometer(n int) (r int) {
	digs := "012356789"
	for i, v := range Reverse(fmt.Sprintf("%d", n)) {
		r += strings.Index(digs, string(v)) * PowInt(9, i)
	}
	return
}

File: Python/5 kyu/Simple Fun #178 Faulty Odometer/Kata.py
# def faulty_odometer(num):
#     digs = '012356789'
#     result = 0
#     for i, n in enumerate(str(num)[::-1]):
#         result += digs.index(n) * 9 ** i
#     return result

def faulty_odometer(n):
    if n == 0:
        return 0
    else:
        d = n % 10
        if d > 4:
            d -= 1
        return d + 9 * faulty_odometer(n//10)
